File: langs/qr.m
// deps: macos + coreimage
#import <Foundation/Foundation.h>
#import <CoreImage/CoreImage.h>

int main() {
  @autoreleasepool {
    NSData *data = [@"https://example.com" dataUsingEncoding:NSUTF8StringEncoding];
    CIFilter *f = [CIFilter filterWithName:@"CIQRCodeGenerator"];
    [f setValue:data forKey:@"inputMessage"];
    CIImage *ci = f.outputImage;
    CIContext *ctx = [CIContext contextWithOptions:nil];
    CGImageRef cg = [ctx createCGImage:ci fromRect:ci.extent];
    NSBitmapImageRep *rep = [[NSBitmapImageRep alloc] initWithCGImage:cg];
    NSData *png = [rep representationUsingType:NSPNGFileType properties:@{}];
    [png writeToFile:@"qr.png" atomically:YES];
    NSLog(@"qr.png");
    CGImageRelease(cg);
  }
  return 0;
}

File: langs/qr.swift
// deps: macos + coreimage
import Foundation
import CoreImage
import CoreImage.CIFilterBuiltins
import AppKit

let data = "https://example.com".data(using: .utf8)!
let filter = CIFilter.qrCodeGenerator()
filter.message = data
let ciImage = filter.outputImage!
let rep = NSCIImageRep(ciImage: ciImage)
let nsImage = NSImage(size: rep.size)
nsImage.addRepresentation(rep)
let png = NSBitmapImageRep(data: nsImage.tiffRepresentation!)!
let dataOut = png.representation(using: .png, properties: [:])!
try dataOut.write(to: URL(fileURLWithPath: "qr.png"))
print("qr.png")
